components/relevancy/src/lib.rs

```diff
@@ -159,8 +161,76 @@ impl RelevancyStore {
     /// its likelihood of a negative outcome.
     #[handle_error(Error)]
     pub fn bandit_update(&self, bandit: String, arm: String, selected: bool) -> ApiResult<()> {
+        let mut cache = self.cache.lock();
+
+        cache
+            .clear(&bandit, &arm)
+            .expect("failed to clear cached distribution");
+
         self.db
             .read_write(|dao| dao.update_bandit_arm_data(&bandit, &arm, selected))?;
+
+        Ok(())
+    }
+}
+
+pub struct BanditCache {
+    cache: HashMap<(String, String), (usize, usize)>,
+}
+
+impl Default for BanditCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BanditCache {
+    /// Creates a new, empty `BanditCache`.
+    ///
+    /// The cache is initialized as an empty `HashMap` and is used to store
+    /// precomputed Beta distribution parameters for faster access during
+    /// Thompson Sampling operations.
+    pub fn new() -> Self {
+        BanditCache {
+            cache: HashMap::new(),
+        }
+    }
+
+    /// Retrieves the Beta distribution parameters for a given bandit and arm.
+    ///
+    /// If the parameters for the specified `bandit` and `arm` are already cached,
+    /// they are returned directly. Otherwise, the parameters are fetched from
+    /// the database, added to the cache, and then returned.
+    pub fn get_beta_distribution(
+        &mut self,
+        bandit: &str,
+        arm: &str,
+        db: &RelevancyDb,
+    ) -> Result<(usize, usize)> {
+        let key = (bandit.to_string(), arm.to_string());
+
+        // Check if the distribution is already cached
+        if let Some(&params) = self.cache.get(&key) {
+            return Ok(params);
+        }
+
+        let params = db.read(|dao| dao.retrieve_bandit_arm_beta_distribution(bandit, arm))?;
+
+        // Cache the retrieved parameters for future use
+        self.cache.insert(key, params);
+
+        Ok(params)
+    }
+
+    /// Clears the cached Beta distribution parameters for a given bandit and arm.
+    ///
+    /// This removes the cached values for the specified `bandit` and `arm` from the cache.
+    /// Use this method if the cached parameters are no longer valid or need to be refreshed.
+    pub fn clear(&mut self, bandit: &str, arm: &str) -> Result<()> {
```

Comment:
This function never fails, so we can remove the `Result` here and return nothing.